elsa-typed-builtin.el

```diff
@@ -175,36 +170,37 @@
 
 ;; File: fns.c
 ;; (put 'identity 'elsa-type (elsa-make-type ))
-;; (put 'random 'elsa-type (elsa-make-type ))
+(put 'random 'elsa-type (elsa-make-type Int | T | String -> Int))
 (put 'length 'elsa-type (elsa-make-type Sequence? -> Int))
 (put 'safe-length 'elsa-type (elsa-make-type Mixed -> Int))
-;; (put 'string-bytes 'elsa-type (elsa-make-type ))
-;; (put 'string-equal 'elsa-type (elsa-make-type ))
+(put 'string-bytes 'elsa-type (elsa-make-type String -> Int))
+(put 'string-equal 'elsa-type (elsa-make-type String | Symbol -> String | Symbol -> Bool))
 (put 'compare-strings 'elsa-type (elsa-make-type String -> (Int | Nil) -> (Int | Nil) -> String -> (Int | Nil) -> (Int | Nil) -> Bool? -> (Int | T)))
-;; (put 'string-lessp 'elsa-type (elsa-make-type ))
-;; (put 'string-collate-lessp 'elsa-type (elsa-make-type ))
-;; (put 'string-collate-equalp 'elsa-type (elsa-make-type ))
-;; (put 'append 'elsa-type (elsa-make-type ))
-;; (put 'concat 'elsa-type (elsa-make-type ))
-;; (put 'vconcat 'elsa-type (elsa-make-type ))
+(put 'string-lessp 'elsa-type (elsa-make-type String | Symbol -> String | Symbol -> Bool))
+(put 'string-collate-lessp 'elsa-type (elsa-make-type String | Symbol -> String | Symbol -> String? -> Bool? -> Bool))
+(put 'string-collate-equalp 'elsa-type (elsa-make-type String | Symbol -> String | Symbol -> String? -> Bool? -> Bool))
+(put 'append 'elsa-type (elsa-make-type Sequence... -> List))
+(put 'concat 'elsa-type (elsa-make-type Variadic (String | List Int | Vector Int) -> String))
+(put 'vconcat 'elsa-type (elsa-make-type Sequence... -> Vector))
 ;; (put 'copy-sequence 'elsa-type (elsa-make-type ))
 ;; (put 'string-make-multibyte 'elsa-type (elsa-make-type ))
 ;; (put 'string-make-unibyte 'elsa-type (elsa-make-type ))
 ;; (put 'string-as-unibyte 'elsa-type (elsa-make-type ))
 ;; (put 'string-as-multibyte 'elsa-type (elsa-make-type ))
 ;; (put 'string-to-multibyte 'elsa-type (elsa-make-type ))
 ;; (put 'string-to-unibyte 'elsa-type (elsa-make-type ))
-;; (put 'copy-alist 'elsa-type (elsa-make-type ))
-;; (put 'substring 'elsa-type (elsa-make-type ))
-;; (put 'substring-no-properties 'elsa-type (elsa-make-type ))
+(put 'copy-alist 'elsa-type (elsa-make-type List[Cons Mixed Mixed] -> List[Cons Mixed Mixed]))
```

Comment:
Maybe this is an issue: `List[Cons...]` is this what you wanted? It's a list of lists.

Reply:
When you write `List [a]` this is a list of lists of a.  A list of things is either `List (a)` or `[a]`.  It's a bit confusing.

Your second observation is correct, that is exactly how a list is defined in elisp: either a nil, or a cons... except that the last cell must be nil but Elsa does not enforce that yet.

Reply:
```elisp
(elsa-type-describe (elsa-make-type List [Cons])) => "[[Cons Mixed Mixed]]"
```

this is how you can get a "canonical" readable type expression (readable as in can be passed to `elsa-make-type` to give you back the input).

Reply:
Yea, if it's an alist.  `Cons` is short for `Cons Mixed Mixed`, same as `List` is `List Mixed` and `Vector` is `Vector Mixed`.

If the alist is special, for example from symbols to strings you can write more specific version `[Cons Symbol String]`.